modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/ScanNode.java

```diff
@@ -58,8 +60,20 @@ public ScanNode(ExecutionContext<Row> ctx, RelDataType rowType, Iterable<Row> sr
 
         requested = rowsCnt;
 
-        if (!inLoop)
-            context().execute(this::push, this::onError);
+        if (!inLoop) {
+            if (firstReq) {
```

Comment:
seems like stub as for me, wins here you can loose on real concurrent processing.

Reply:
This is simply not true, this is always win (context switching is not for free)

Reply:
Here we just process the first batch together with the query start request, it should not affect somehow concurrent processing. As an alternative, we can create a separate task for sending `QueryStartResponse` after initiating `push` task. WDYT about it?

Reply:
I`m not debate about context switching, we on the same side here, i talking about concurrent overall throughput here.

Reply:
Overall throughput can't be changed by this change. The only difference here: now there is no gap between QueryStartRequest and the first batch from data source processing. But batch size is relatively small to affect response time for other queries. If you still think that it's a problem, WDYT about sending QueryStartResponse in a separate task after the first push task?

Reply:
I really don`t know i just suppose, seems we need some kind of yardstick tests here, if you have no possibility to run them let's leave it as it is.